File: backend/app/schemas/datasource.py
from pydantic import BaseModel, Field


class DatasourceCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    type: str = Field(..., pattern=r"^(mysql|postgresql|mssql|oracle)$")
    host: str = Field(..., min_length=1, max_length=255)
    port: int = Field(..., ge=1, le=65535)
    database: str = Field(..., min_length=1, max_length=255)
    username: str = Field(..., min_length=1, max_length=255)
    password: str = Field(..., min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=500)
    table_blacklist: str | None = "[]"
    column_blacklist: str | None = "[]"


class DatasourceUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    host: str | None = Field(default=None, max_length=255)
    port: int | None = Field(default=None, ge=1, le=65535)
    database: str | None = Field(default=None, max_length=255)
    username: str | None = Field(default=None, max_length=255)
    password: str | None = Field(default=None, max_length=500)
    is_active: bool | None = None
    description: str | None = Field(default=None, max_length=500)
    table_blacklist: str | None = None
    column_blacklist: str | None = None


class DatasourceResponse(BaseModel):
    id: int
    name: str
    type: str
    host: str
    port: int
    database: str
    username: str
    is_active: bool
    description: str | None = None
    table_blacklist: str | None = "[]"
    column_blacklist: str | None = "[]"

    model_config = {"from_attributes": True}

File: frontend/src/api/datasource.ts
import request from './index'

export interface Datasource {
  id: number
  name: string
  type: string
  host: string
  port: number
  database: string
  username: string
  description?: string
  is_active: boolean
  table_blacklist?: string
  column_blacklist?: string
  created_at: string
}

export interface DatasourceCreate {
  name: string
  type: string
  host: string
  port: number
  database: string
  username: string
  password: string
  description?: string
  table_blacklist?: string
  column_blacklist?: string
}

export function listDatasources(): Promise<Datasource[]> {
  return request.get('/datasources')
}

export function getDatasource(id: number): Promise<Datasource> {
  return request.get(`/datasources/${id}`)
}

export function createDatasource(data: DatasourceCreate): Promise<Datasource> {
  return request.post('/datasources', data)
}

export function updateDatasource(id: number, data: Partial<DatasourceCreate>): Promise<Datasource> {
  return request.put(`/datasources/${id}`, data)
}

export function deleteDatasource(id: number): Promise<void> {
  return request.delete(`/datasources/${id}`)
}

export function testDatasourceConnection(id: number): Promise<{ success: boolean; message: string }> {
  return request.post(`/datasources/${id}/test`)
}
